### PR TITLE
cross validated results

### PR DESCRIPTION
**Pull request by YAY-C**
- [x] added save-file for results; added sub-selection parameters
- [x] added cosine kernel
- [x] added score (r2) evaluation _(maybe not needed since individual predictions are saved)_
- [x] add explicit random-state to regression
- [x] What about the `--debug` parameter ? 
- [x] add KWARG to save predictions for each seed
- [x]  add test for `regression`, `hyperparameters`

### CODE REVIEW

**Comment by YAY-C** [comment]
I realize now that we have no tests for the regression/prediction scripts!
_how would you test them ? on a toy-example or a small dataset included as example ?_

File: qstack/regression/cross_validate_results.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+import sys
+import numpy as np
+import scipy
+from sklearn.model_selection import train_test_split, KFold
+from qstack.regression.kernel_utils import get_kernel, defaults, ParseKwargs
+from qstack.regression.hyperparameters import hyperparameters
+from qstack.regression.regression import regression
+from qstack.tools import correct_num_threads
+import qstack.spahm.rho.utils as utt
```
Comment (YAY-C) [inline qstack/regression/cross_validate_results.py:11]: - [ ] change name to `utils`

File: qstack/regression/cross_validate_results.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+
+
+def cv(X, y,
```
Comment (YAY-C) [inline qstack/regression/cross_validate_results.py:15]: - [ ] change name _more explicit_

File: qstack/regression/cross_validate_results.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+           sigmaarr=defaults.sigmaarr, etaarr=defaults.etaarr, gkernel=defaults.gkernel, gdict=defaults.gdict,
+           akernel=defaults.kernel, test_size=defaults.test_size, train_size=defaults.train_size, splits=defaults.splits,
+           printlevel=0, adaptive=False, read_kernel=False, ipywidget=None, n_rep=defaults.n_rep, save=False, preffix='uknown', save_pred=False):
```
Comment (YAY-C) [inline qstack/regression/cross_validate_results.py:18]: - [ ] remove `ipywidget`

File: qstack/regression/cross_validate_results.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    hyper_runs = []
+    lc_runs = []
+    seeds = [123, 1, 2, 66, 666, 18, 9, 1996, 26,  3, 17]
```
Comment (YAY-C) [inline qstack/regression/cross_validate_results.py:21]: - [ ] parametrize w\ respect to number of runs

File: qstack/regression/cross_validate_results.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    if save_pred: predictions_n = []
+    bar = 1 if printlevel == 0 else 0
+    if bar > 0:
+        progress = utt.add_progressbar(max_value=n_rep)
```
Comment (YAY-C) [inline qstack/regression/cross_validate_results.py:25]: - [ ] make consistent and change to `tqdm`

File: qstack/regression/cross_validate_results.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    for seed,n in zip(seeds, range(n_rep)):
+        error = hyperparameters(X, y, read_kernel=False, sigma=sigmaarr, eta=etaarr, akernel=akernel, test_size=test_size, splits=splits, printlevel=printlevel, adaptive=adaptive, debug=seed)
```
Comment (YAY-C) [inline qstack/regression/cross_validate_results.py:27]: - [ ] rename `debug`

File: qstack/regression/cross_validate_results.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        mae, stdev, eta, sigma = zip(*error)
+        maes_all = regression(X, y, read_kernel=False, sigma=sigma[-1], eta=eta[-1], akernel=akernel, test_size=test_size, train_size=train_size, n_rep=1, debug=seed, save_pred=save_pred)
```
Comment (YAY-C) [inline qstack/regression/cross_validate_results.py:29]: - [ ] rename `debug`

File: qstack/regression/cross_validate_results.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    parser.add_argument('--ll',   action='store_true', dest='ll',       default=False,  help='if correct for the numper of threads')
+    parser.add_argument('--save',   action='store_true', dest='save_all',       default=False,  help='if saving intermediate results in .npy file')
+    parser.add_argument('--ada',  action='store_true', dest='adaptive', default=False,  help='if adapt sigma')
+    parser.add_argument('--save-pred',  action='store_true', dest='save_pred', default=False,  help='if save test-set prediction')
+    parser.add_argument('--readkernel', action='store_true', dest='readk', default=False,  help='if X is kernel')
+    parser.add_argument('--name',      type=str,   dest='nameout',       required=True, help='the name of the output file')
+    parser.add_argument('--select',      type=str,   dest='f_select',       required=False, help='a txt file containing the indices of the selected representations')
+    parser.add_argument('--sub',      action="store_true",   dest='sub_test',       required=False, help='run fast test (10 sub-data points)')
+    args = parser.parse_args()
```
Comment (YAY-C) [inline qstack/regression/cross_validate_results.py:81]: - [ ] remove `--sub` option
- [ ] add `--sparse` option

File: qstack/regression/hyperparameters.py [hunk 9]
```diff
@@ -11,7 +11,7 @@
 def hyperparameters(X, y,
            sigma=defaults.sigmaarr, eta=defaults.etaarr, gkernel=defaults.gkernel, gdict=defaults.gdict,
            akernel=defaults.kernel, test_size=defaults.test_size, splits=defaults.splits,
-           printlevel=0, adaptive=False, read_kernel=False, sparse=None):
+           printlevel=0, adaptive=False, read_kernel=False, sparse=None, debug=0):
```
Comment (YAY-C) [inline qstack/regression/hyperparameters.py:14]: - [ ] rename `debug`

File: qstack/regression/regression.py [hunk 10]
```diff
@@ -3,16 +3,16 @@
 import numpy as np
 import scipy
 from sklearn.model_selection import train_test_split
+from sklearn.metrics import r2_score
```
Comment (YAY-C) [inline qstack/regression/regression.py:6]: **maybe not needed**
_make separate utility to computes various metrics (**new issue ?**) ?_

File: qstack/regression/regression.py [hunk 11]
```diff
@@ -3,16 +3,16 @@
...
 from qstack.tools import correct_num_threads
 from qstack.mathutils.fps import do_fps
 
-
 def regression(X, y, read_kernel=False, sigma=defaults.sigma, eta=defaults.eta,
                akernel=defaults.kernel, gkernel=defaults.gkernel, gdict=defaults.gdict,
                test_size=defaults.test_size, train_size=defaults.train_size, n_rep=defaults.n_rep,
                random_state=defaults.random_state,
-               sparse=None, debug=False):
+               sparse=None, debug=False, save_pred=False):
```
Comment (YAY-C) [inline qstack/regression/regression.py:15]: - [ ] rename `debug`

File: qstack/regression/regression.py [hunk 12]
```diff
@@ -84,16 +85,24 @@ def main():
...
     X = np.load(args.repr)
     y = np.loadtxt(args.prop)
+    if args.f_select != None:
+        selected = np.loadtxt(args.f_select, dtype=int)
+        X = X[selected]
+        y = y[selected]
     maes_all = regression(X, y, read_kernel=args.readk, sigma=args.sigma, eta=args.eta, akernel=args.akernel,
                           test_size=args.test_size, train_size=args.train_size, n_rep=args.splits, sparse=args.sparse,
-                          debug=args.debug)
+                          debug=args.debug) ##TODO: add args.random_state and what about debug (obsolete?)
```
Comment (YAY-C) [inline qstack/regression/regression.py:101]: - [ ] add `args.random_state` and rename `debug`

File: qstack/regression/regression.py [hunk 13]
```diff
@@ -84,16 +85,24 @@ def main():
     parser.add_argument('--readkernel',    action='store_true', dest='readk', default=False,  help='if X is kernel')
     parser.add_argument('--sparse',        type=int, dest='sparse', default=None,  help='regression basis size for sparse learning')
     parser.add_argument('--random_state',  type=int, dest='random_state', default=defaults.random_state,  help='random state for test / train splitting')
+    parser.add_argument('--select',        type=str,   dest='f_select',       required=False, help='a txt file containing the indices of the selected representations')
+    parser.add_argument('--name',          type=str,   dest='nameout',     required=True,    help='the name of the output file containting the LC data (.txt).')
```
Comment (YAY-C) [inline qstack/regression/regression.py:89]: - [ ] make **not required** and `None` _(dont' save is default ?)_

File: qstack/regression/regression.py (hunk 10 above)
Comment (YAY-C) [reply qstack/regression/regression.py:6]: Actually, it might also be relevant for the project with Puck and the Reiher group.
_I'll make an issue/pr draft._

**Comment by briling** [comment]
> how would you test them ? on a toy-example or a small dataset included as example ?

just give them any X and Y files. We have something in https://github.com/lcmd-epfl/spahm and https://github.com/lcmd-epfl/spahm-rho

File: qstack/regression/hyperparameters.py [hunk 14]
```diff
@@ -127,16 +127,25 @@ def main():
     parser.add_argument('--ada',  action='store_true', dest='adaptive', default=False,  help='if adapt sigma')
     parser.add_argument('--readkernel', action='store_true', dest='readk', default=False,  help='if X is kernel')
     parser.add_argument('--sparse',     type=int, dest='sparse', default=None,  help='regression basis size for sparse learning')
+    parser.add_argument('--name',      type=str,   dest='nameout',       required=False, default=None,  help='the name of the output file')
+    parser.add_argument('--select',      type=str,   dest='f_select',       required=False, help='a txt file containing the indices of the selected representations')
```
[the remark this answers is not in the document]
Comment (YAY-C) [reply qstack/regression/hyperparameters.py:131]: Yes, they can if they use the function directly but not if they use `main()`
Otherwise they need to regenerated the `.npy` arrays (passed to CLI) with the restricted selection.

This way you just have one `.npy` array and then as many `.txt` lists (containing indices) as you want to try out various selection.

**Review by YAY-C — COMMENTED** [review]
- [x] add a `seed0` parameter to control the content of the array `seeds`

File: qstack/regression/cross_validate_results.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+            sigmaarr=defaults.sigmaarr, etaarr=defaults.etaarr, gkernel=defaults.gkernel,
+            gdict=defaults.gdict, akernel=defaults.kernel, test_size=defaults.test_size,
+            train_size=defaults.train_size, splits=defaults.splits, printlevel=0,
+            adaptive=False, read_kernel=False, n_rep=defaults.n_rep, save=False,
+            preffix='unknown', save_pred=False, progress=False, sparse=None,
+            debug=None):
+
+    hyper_runs = []
+    lc_runs = []
+    seeds = np.arange(n_rep)
```
[the remark this answers is not in the document]
Comment (briling) [reply qstack/regression/cross_validate_results.py:25]: yeah, either as `seed0` or the list of seeds

[the remark this answers is not in the document]
Comment (YAY-C) [reply qstack/regression/cross_validate_results.py:25]: :+1:

**Comment by briling** [comment]
do we need the xyz files?

**Comment by YAY-C** [comment]
> do we need the xyz files?

Not necessarily, but I thought we might need them for upcomingx tests.
_Do you think it makes more sense to remove them ?_

File: qstack/regression/hyperparameters.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (YAY-C) [reply qstack/regression/hyperparameters.py:131]: Actually, now that I had to use it I remember, the main point of this option is to be able to restrict the initial dataset. Typically, if I want to truncate or slice the initial dataset according to some threshold on the property (to change/smoothen its distribution) I just list the indices instead of producing a new array of representations AND a new property file. But then the rest should run normally/random.
_Is it too far fetch? maybe this type of stuff should be dealt by some home-made script calling the function instead?_

[the remark this answers is not in the document]
Comment (YAY-C) [reply qstack/regression/hyperparameters.py:131]: Also, though 
> then it would make sense to have txt files for train/validation/test, what do you think?    

this would make sense for hyperparameter and regression, it wouldn't make sense for cross_validate_results, because then you would need to specify train/val/test sets for all repetitions (normally over seeds), since the point is to vary the test set.

[the remark this answers is not in the document]
Comment (briling) [reply qstack/regression/hyperparameters.py:131]: > Is it too far fetch? maybe this type of stuff should be dealt by some home-made script calling the function instead?

[the remark this answers is not in the document]
Comment (briling) [reply qstack/regression/hyperparameters.py:131]: > this would make sense for hyperparameter and regression, it wouldn't make sense for cross_validate_results

**Comment by briling** [comment]
> > do we need the xyz files?
> 
> Not necessarily, but I thought we might need them for upcoming tests. _Do you think it makes more sense to remove them ?_

Could you please add a test that generates the representation (used in this test) from these xyz files? so we remember they're related

File: qstack/regression/hyperparameters.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (YAY-C) [reply qstack/regression/hyperparameters.py:131]: Then I will remove the `--select` option from all three scripts.
And I will add a `--selection` option and document it in the `parser` help!

**Comment by YAY-C** [comment]
> > > do we need the xyz files?
> > 
> > 
> > Not necessarily, but I thought we might need them for upcoming tests. _Do you think it makes more sense to remove them ?_
> 
> Could you please add a test that generates the representation (used in this test) from these xyz files? so we remember they're related

- [ ] add test-data generation script

**Review by YAY-C — COMMENTED** [review]
Do you think we need a CLI arg for the selection thing ?

_I will also add the docs so it's done!_

- [x] add documentation
